irispy/new_sji.py

```diff
@@ -0,0 +1,179 @@
+'''
+This module provides movie tools for level 2 IRIS SJI fits file
+'''
+
+from datetime import timedelta
+
+from ndcube import NDCube
+import numpy as np
+from astropy.io import fits
+import astropy.units as u
+from astropy.wcs import WCS
+from sunpy.time import parse_time
+from sunpy import config
+from irispy import iris_tools
+
+__all__ = ['SJICube']
+
+TIME_FORMAT = config.get("general", "time_format")
+
+# the following value is only appropriate for byte scaled images
+BAD_PIXEL_VALUE = -200
+# the following value is only appropriate for unscaled images
+BAD_PIXEL_VALUE_UNSCALED = -32768
+
+
+class SJICube(NDCube):
+    """
+    SJICube
+
+    Class representing SJI Images described by a single WCS
+
+    Parameters
+    ----------
+    data: `numpy.ndarray`
+        The array holding the actual data in this object.
+
+    wcs: `ndcube.wcs.wcs.WCS`
+        The WCS object containing the axes' information
+
+    unit : `astropy.unit.Unit` or `str`
+        Unit for the dataset.
+        Strings that can be converted to a Unit are allowed.
+
+    meta : dict-like object
+        Additional meta information about the dataset.
+
+    uncertainty : any type, optional
+        Uncertainty in the dataset. Should have an attribute uncertainty_type
+        that defines what kind of uncertainty is stored, for example "std"
+        for standard deviation or "var" for variance. A metaclass defining
+        such an interface is NDUncertainty - but isn’t mandatory. If the
+        uncertainty has no such attribute the uncertainty is stored as
+        UnknownUncertainty.
+        Defaults to None.
+
+    mask : any type, optional
+        Mask for the dataset. Masks should follow the numpy convention
+        that valid data points are marked by False and invalid ones with True.
+        Defaults to None.
+
+    extra_coords : iterable of `tuple`s, each with three entries
+        (`str`, `int`, `astropy.units.quantity` or array-like)
+        Gives the name, axis of data, and values of coordinates of a data axis
+        not included in the WCS object.
+
+    copy : `bool`, optional
+        Indicates whether to save the arguments as copy. True copies every
+        attribute before saving it while False tries to save every parameter
+        as reference. Note however that it is not always possible to save the
+        input as reference.
+        Default is False.
+
+    Examples
+    --------
+    >>> from irispy import sji
+    >>> from irispy.data import sample
+    >>> sji = read_iris_sji_level2_fits(sample.SJI_CUBE_1400)
+    """
```

Comment:
If we are introducing the option to memory map (i.e. scale or not scale) the data, we should add a flag to state that.  So say, ```SJICube.scaled = True``` if memmap is ```False```, ```SJICube.scaled = False``` if memmap is ```True```, or ```SJICube.scaled = None``` if not specified.

This will require the ```SJICube.__init__``` to be put back as we are now doing something more than just the calling the ```NDCube.__init__```.  So the ```SJICube.__init__``` would look something like:
```
def __init__(self, data, wcs, uncertainty=None, unit=None, meta=None, mask=None, 
             extra_coords=None, copy=False, missing_axis=None, scaled=None):
    # Set whether SJI data is scaled or not.
    self.scaled = scaled
    # Call NDCube init
    super............
```
Then ```scaled``` must be set to ```True``` or ```False``` in ```read_iris_sji_level2_fits``` when it defines the ```SJICube``` object based on the value of the ```memmap``` keyword argument (kwarg).

Having the ```scaled``` attribute will enable a user to know how to interpret the data numbers even without knowing whether ```memmap``` was ```True``` or ```False``` during initialization.

We should also add an ```SJI_UNSCALED``` unit to the DN dictionary in ```iris_tools``` (https://github.com/sunpy/irispy/blob/master/irispy/iris_tools.py#L27) which is just the ```DN_UNIT["SJI"]/SJI_SCALE_FACTOR```.    (```SJI_SCALE_FACTOR``` should also be defined near the top of ```iris_tools.py```.)  Then based on whether ```memmap``` is ```True``` or ```False``` the ```unit``` of ```SJICube``` can be set to ```DN_UNIT["SJI"]``` or ```DN_UNIT["SJI_UNSCALED"]``` by ```read_iris_sji_level2_fits```.

Reply:
Perhaps the ```scaled``` argument should be a required argument, like ```data``` and ```wcs```.  Or perhaps it should default to ```True```.  Either of these ways will mean it will always have a Boolean value and will never be ```None```.  That way any infrastructure depending on it, e.g. ```unit``` will always be valid.  I'll give this is a little thought.

Reply:
How can I define `SJI_SCALE_FACTOR` on `iris_tools.py` ?

`scaled` has been added on reading function and SJI class.

Reply:
Firstly, could @tiagopereira chime in here and confirm whether the scaling is always the same for SJI data?  Depending on that answer, we can decide how to define it.

Reply:
ok I push it now

Reply:
I couldn't find an authoritative source on this, but for all the files I tested, the "masked" values in unscaled FITS units are always -32768. Makes sense since it is the lowest signed 16-bit integer. Also for all files (SJI and SG), `BSCALE` and `BZERO` seem to be the same, but  these we can always read from the header. I think it is safe to assume that the bad pixel value is always -32768.

Reply:
OK in that case the definition of ```BAD_PIXEL_VALUE_UNSCALED``` in this PR is fine :)

If we aren't comfortable just assuming what the ```BSCALE``` and ```BZERO``` we shall have to think how best to define the equivalency between the scaled SJI data unit, and the unscaled one.  I'll give this some thought.